### PR TITLE
[ISSUE #3624]fix DefaultMQPushConsumerTest.testPullMessage_ExceptionO…

### PR DESCRIPTION
**Pull request by panzhi33**
…ccursWhenComputePullFromWhere throws RejectedExecution

**Make sure set the target branch to `develop`**

## What is the purpose of the change

XXXXX

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java [hunk 1]
```diff
@@ -74,14 +69,39 @@ public void scanExpiredRequest() {
...
+                        log.error("scan RequestFutureTable exception", e);
+                    }
+                }
+            }, 1000 * 3, 1000, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    public synchronized void shutdown() {
+        if (this.producerNum.decrementAndGet() == 0) {
+            this.getScheduledExecutorService().shutdown();
```
Comment (areyouok) [inline client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:89]: set the scheduledExecutorService to null?

Comment (panzhi33) [reply client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java]: ok

Comment (areyouok) [reply client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java]: 1. shutdown 
2. set to null

**Review by caigy — APPROVED** [review]

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java (hunk 1 above)
Comment (panzhi33) [reply client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java]: ok

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java [hunk 2]
```diff
@@ -37,19 +38,13 @@
     private static final RequestFutureHolder INSTANCE = new RequestFutureHolder();
     private ConcurrentHashMap<String, RequestResponseFuture> requestFutureTable = new ConcurrentHashMap<String, RequestResponseFuture>();
     private final AtomicInteger producerNum = new AtomicInteger(0);
```
Comment (areyouok) [inline client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:40]: there is no need use AtomicInteger.

I suggest change to HashSet<DefaultMQProducerImpl> and startScheduledTask/shutdown support re-entry. 

eg: called DefaultMQProducerImpl.shutdown got exception, and the use may call shutdown again.

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java [hunk 3]
```diff
@@ -74,14 +69,40 @@ public void scanExpiredRequest() {
...
     }
 
-    public AtomicInteger getProducerNum() {
-        return producerNum;
+    private RequestFutureHolder() {
     }
 
-    public ScheduledExecutorService getScheduledExecutorService() {
+    private ScheduledExecutorService getScheduledExecutorService() {
+        if (null == scheduledExecutorService || scheduledExecutorService.isShutdown()) {
```
Comment (areyouok) [inline client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:98]: no need to check isShutdown

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java [hunk 4]
```diff
@@ -74,14 +69,40 @@ public void scanExpiredRequest() {
...
+                        log.error("scan RequestFutureTable exception", e);
+                    }
+                }
+            }, 1000 * 3, 1000, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    public synchronized void shutdown() {
+        if (this.producerNum.decrementAndGet() == 0) {
+            this.scheduledExecutorService.shutdown();
```
Comment (areyouok) [inline client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:89]: check null

**Review by areyouok — CHANGES_REQUESTED** [review]

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java (hunk 2 above)
Comment (panzhi33) [reply client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:40]: Don't understand what you mean

Comment (panzhi33) [reply client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:40]: like this
if(set.size > 0 && state != running) {
   startScheduledTask
}

if(set.size<=0 && state != shutdown){
  shutdown 
}

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java [hunk 5]
```diff
@@ -74,14 +73,47 @@ public void scanExpiredRequest() {
...
+        }
     }
 
-    public ScheduledExecutorService getScheduledExecutorService() {
+    private RequestFutureHolder() {
+    }
+
+    private ScheduledExecutorService getScheduledExecutorService() {
+        if (null == scheduledExecutorService) {
+            scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactory() {
```
Comment (Git-Yang) [inline client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:110]: Maybe better to replace ThreadFactory with ThreadFactoryImpl.

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java [hunk 6]
```diff
@@ -74,14 +73,42 @@ public void scanExpiredRequest() {
...
+            this.scheduledExecutorService = null;
+            this.serviceState = ServiceState.SHUTDOWN_ALREADY;
+        }
     }
 
-    public ScheduledExecutorService getScheduledExecutorService() {
+    private RequestFutureHolder() {
+    }
+
+    private ScheduledExecutorService getScheduledExecutorService() {
```
Comment (areyouok) [inline client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:108]: this method should inline

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java [hunk 7]
```diff
@@ -74,14 +73,42 @@ public void scanExpiredRequest() {
         }
     }
 
-    private RequestFutureHolder() {
+    public synchronized void startScheduledTask(DefaultMQProducerImpl producer) {
+        this.producerSet.add(producer);
+        if (this.producerSet.size() >= 1 && this.serviceState != ServiceState.RUNNING) {
```
Comment (areyouok) [inline client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:78]: no need check producerSet.size()

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java [hunk 8]
```diff
@@ -17,39 +17,38 @@
...
-    private final ScheduledExecutorService scheduledExecutorService = Executors
-        .newSingleThreadScheduledExecutor(new ThreadFactory() {
-            @Override
-            public Thread newThread(Runnable r) {
-                return new Thread(r, "RequestHouseKeepingService");
-            }
-        });
+    private final Set<DefaultMQProducerImpl> producerSet = new HashSet<>();
+    private ScheduledExecutorService scheduledExecutorService = null;
+    private ServiceState serviceState = ServiceState.CREATE_JUST;
```
Comment (areyouok) [inline client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:45]: this serviceState is duplicate. we can use scheduledExecutorService==null to check

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java (hunk 4 above)
Comment (areyouok) [reply client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java:99]: may thutdown() throws exception? 

the below code maybe better?
```
ScheduledExecutorService ses = this.scheduledExecutorService;
this.scheduledExecutorService = null;
ses.shutdown();
```

**Review by areyouok — CHANGES_REQUESTED** [review]

File: client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java (hunk 4 above)
Comment (panzhi33) [reply client/src/main/java/org/apache/rocketmq/client/producer/RequestFutureHolder.java]: yes，it would be better

**Review by areyouok — APPROVED** [review]

**Pull request merged**